gemrb/core/Audio/Playback.cpp

```diff
@@ -0,0 +1,184 @@
+	: source(std::move(source)), length(length), height(height)
+{}
+
+time_t PlaybackHandle::GetLength() const
+{
+	return length;
+}
+
+bool PlaybackHandle::IsPlaying() const
+{
+	if (source) {
+		return !source->HasFinishedPlaying();
+	}
+
+	return false;
+}
+
+void PlaybackHandle::SetPosition(const Point& p)
+{
+	if (source) {
+		source->SetPosition({p, height});
+	}
+}
+
+void PlaybackHandle::StopLooping()
+{
+	if (source) {
+		source->StopLooping();
+	}
+}
+
+void PlaybackHandle::Stop()
+{
+	if (source) {
+		source->Stop();
+		source.reset();
+	}
+}
+
+AudioPlayback::AudioPlayback(const std::vector<ResRef>& defaultSounds) : defaultSounds(defaultSounds)
+{}
+
+Holder<PlaybackHandle> AudioPlayback::Play(StringView resource, const AudioPlaybackConfig& config)
+{
+	Housekeeping();
+
+	if (resource.empty()) {
+		return {};
+	}
+
+	auto cacheEntry = GetBuffer(resource, config);
+	if (!cacheEntry.handle) {
+		return {};
+	}
+
+	auto source = core->GetAudioDrv()->CreatePlaybackSource(config);
+	source->Enqueue(std::move(cacheEntry.handle));
+
+	activeSources.insert(source);
+
+	return MakeHolder<PlaybackHandle>(std::move(source), cacheEntry.length, config.position.z);
+}
+
+Holder<PlaybackHandle> AudioPlayback::PlayDefaultSound(size_t index, SFXChannel channel)
+{
+	return PlayDefaultSound(index, core->GetAudioSettings().ConfigPresetScreenAction(channel));
+}
+
+Holder<PlaybackHandle> AudioPlayback::PlayDefaultSound(size_t index, const AudioPlaybackConfig& config)
+{
+	Housekeeping();
+
+	if (index >= defaultSounds.size()) {
+		return {};
+	}
+
+	auto resource = defaultSounds[index];
+	auto cacheEntry = GetBuffer(resource, config);
+	if (!cacheEntry.handle) {
+		return {};
+	}
+
+	auto source = core->GetAudioDrv()->CreatePlaybackSource(config);
+	source->Enqueue(std::move(cacheEntry.handle));
+
+	activeSources.insert(source);
+
+	return MakeHolder<PlaybackHandle>(std::move(source), cacheEntry.length, config.position.z);
+}
+
+time_t AudioPlayback::PlaySpeech(StringView resource, const AudioPlaybackConfig& config, bool interrupt)
+{
+	Housekeeping();
+
+	if (!speech) {
+		speech = core->GetAudioDrv()->CreatePlaybackSource(config, true);
+	}
+
+	auto cacheEntry = GetBuffer(resource, config);
+	if (!cacheEntry.handle) {
+		return -1;
+	}
+
+	if (interrupt) {
+		speech->Stop();
+	}
+
+	speech->Reconfigure(config);
+	speech->Enqueue(std::move(cacheEntry.handle));
+
+	return cacheEntry.length;
+}
+
+void AudioPlayback::StopSpeech()
+{
+	if (speech) {
+		Housekeeping();
+		speech->Stop();
+	}
+}
+
+BufferCacheEntry AudioPlayback::GetBuffer(StringView resource, const AudioPlaybackConfig& config)
+{
+	auto cacheEntry = bufferCache.Lookup(resource);
+	if (cacheEntry) {
+		return *cacheEntry;
+	}
+
+	ResourceHolder<SoundMgr> acm = gamedata->GetResourceHolder<SoundMgr>(resource);
+	if (!acm) {
+		return {};
+	}
+
+	auto length = ((acm->get_length() / acm->get_channels()) * 1000) / acm->get_samplerate();
```

Comment:
It's the same here, perhaps it's worth it to create another function in SoundMgr.